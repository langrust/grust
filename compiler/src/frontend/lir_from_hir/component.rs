use itertools::Itertools;

use crate::{
    common::r#type::Type,
    hir::{identifier_creator::IdentifierCreator, node::Node},
    lir::{
        expression::Expression as LIRExpression,
        item::state_machine::{
            event::{Event, EventElement},
            input::{Input, InputElement},
            state::{init::Init, step::Step, State},
            StateMachine,
        },
    },
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for Node {
    type LIR = StateMachine;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let Node {
            id,
            statements,
            memory,
            contract,
            ..
        } = self;
        let mut identifier_creator = IdentifierCreator::from(vec![]);

        // get node name
        let name = symbol_table.get_name(id).clone();

        // get node inputs
        let mut inputs = symbol_table
            .get_node_inputs(id)
            .into_iter()
            .map(|id| {
                (
                    symbol_table.get_name(*id).clone(),
                    symbol_table.get_type(*id).clone(),
                )
            })
            .collect::<Vec<_>>();

        // create the event structure
        let event = if let Some(event_enum_id) = symbol_table.get_node_event_enum(id) {
            // add the event input
            let event_id = symbol_table.get_node_event(id).unwrap();
            inputs.push((
                symbol_table.get_name(event_id).clone(),
                Type::Enumeration {
                    name: symbol_table.get_name(event_enum_id).clone(),
                    id: event_enum_id,
                },
            ));

            // get enum elements
            let mut elements = symbol_table
                .get_event_enum_elements(event_enum_id)
                .iter()
                .map(|element_id| EventElement::InputEvent {
                    identifier: symbol_table.get_name(*element_id).clone(),
                    r#type: symbol_table.get_type(*element_id).clone(),
                })
                .collect::<Vec<_>>();

            // get generics
            let generics = elements
                .iter_mut()
                .flat_map(|event_elem| match event_elem {
                    EventElement::InputEvent { r#type, .. } => {
                        r#type.get_generics(&mut identifier_creator)
                    }
                    EventElement::NoEvent => vec![],
                })
                .collect::<Vec<_>>();

            Some(Event {
                node_name: name.clone(),
                elements,
                generics,
            })
        } else {
            None
        };

        // get node output type
        let outputs = symbol_table.get_node_outputs(id);
        let mut output_type = {
            let mut types = outputs
                .iter()
                .map(|(_, output_id)| symbol_table.get_type(*output_id).clone())
                .collect::<Vec<_>>();
            if types.len() == 1 {
                types.pop().unwrap()
            } else {
                Type::Tuple(types)
            }
        };

        // get node output expression
        let outputs = symbol_table.get_node_outputs(id);
        let output_expression = {
            let mut identifiers = outputs
                .iter()
                .map(|(_, output_id)| LIRExpression::Identifier {
                    identifier: symbol_table.get_name(*output_id).clone(),
                })
                .collect::<Vec<_>>();
            if identifiers.len() == 1 {
                identifiers.pop().unwrap()
            } else {
                LIRExpression::Tuple {
                    elements: identifiers,
                }
            }
        };

        // collect imports from statements, inputs and output types, memory and contracts
        let mut imports = statements
            .iter()
            .flat_map(|equation| equation.get_imports(symbol_table))
            .unique()
            .collect::<Vec<_>>();
        let mut inputs_type_imports = inputs
            .iter()
            .flat_map(|(_, typing)| typing.get_imports(symbol_table))
            .unique()
            .collect::<Vec<_>>();
        let mut output_type_imports = output_type.get_imports(symbol_table);
        let mut memory_imports = memory.get_imports(symbol_table);
        let mut contract_imports = contract.get_imports(symbol_table);

        // combining all imports and eliminate duplicates
        imports.append(&mut inputs_type_imports);
        imports.append(&mut output_type_imports);
        imports.append(&mut memory_imports);
        imports.append(&mut contract_imports);
        let imports = imports.into_iter().unique().collect::<Vec<_>>();

        // get input's generics: function types in inputs
        let mut generics = inputs
            .iter_mut()
            .flat_map(|(_, typing)| typing.get_generics(&mut identifier_creator))
            .collect::<Vec<_>>();
        let mut output_generics = output_type.get_generics(&mut identifier_creator);
        generics.append(&mut output_generics);

        // get memory/state elements
        let (elements, state_elements_init, state_elements_step) =
            memory.get_state_elements(symbol_table);

        // transform contract
        let contract = contract.lir_from_hir(symbol_table);

        StateMachine {
            name: name.clone(),
            imports,
            input: Input {
                node_name: name.clone(),
                elements: inputs
                    .into_iter()
                    .map(|(identifier, r#type)| InputElement { identifier, r#type })
                    .collect(),
                generics: generics.clone(),
            },
            event,
            state: State {
                node_name: name.clone(),
                elements,
                step: Step {
                    contract,
                    node_name: name.clone(),
                    generics,
                    output_type,
                    body: statements
                        .into_iter()
                        .map(|equation| equation.lir_from_hir(symbol_table))
                        .collect(),
                    state_elements_step,
                    output_expression,
                },
                init: Init {
                    node_name: name,
                    state_elements_init,
                    invariant_initialisation: vec![], // TODO
                },
            },
        }
    }
}
