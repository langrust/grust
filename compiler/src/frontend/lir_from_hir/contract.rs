prelude! {
    hir::contract::Contract,
    lir::contract::Contract as LIRContract,
}

use super::LIRFromHIR;

impl LIRFromHIR for Contract {
    type LIR = LIRContract;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let Contract {
            requires,
            ensures,
            invariant,
        } = self;

        LIRContract {
            requires: requires
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
            ensures: ensures
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
            invariant: invariant
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
        }
    }
}

mod term {
    prelude! {
        hir::contract::{Term, term},
    }

    impl super::LIRFromHIR for Term {
        type LIR = lir::contract::Term;

        fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
            match self.kind {
                term::Kind::Constant { constant } => lir::contract::Term::literal(constant),
                term::Kind::Identifier { id } => {
                    let name = symbol_table.get_name(id);
                    match symbol_table.get_scope(id) {
                        Scope::Input => lir::contract::Term::input(name),
                        Scope::Memory => lir::contract::Term::mem(name),
                        Scope::Output => lir::contract::Term::ident("result"), // todo: this will broke for components with multiple outputs
                        Scope::Local => lir::contract::Term::ident(name),
                    }
                }
                term::Kind::Enumeration {
                    enum_id,
                    element_id,
                } => lir::contract::Term::enumeration(
                    symbol_table.get_name(enum_id).clone(),
                    symbol_table.get_name(element_id).clone(),
                    None,
                ),
                term::Kind::Unary { op, term } => {
                    lir::contract::Term::unop(op, term.lir_from_hir(symbol_table))
                }
                term::Kind::Binary { op, left, right } => lir::contract::Term::binop(
                    op,
                    left.lir_from_hir(symbol_table),
                    right.lir_from_hir(symbol_table),
                ),
                term::Kind::ForAll { id, term } => {
                    let name = symbol_table.get_name(id);
                    let ty = symbol_table.get_type(id).clone();
                    let term = term.lir_from_hir(symbol_table);
                    lir::contract::Term::forall(name, ty, term)
                }
                term::Kind::Implication { left, right } => lir::contract::Term::implication(
                    left.lir_from_hir(symbol_table),
                    right.lir_from_hir(symbol_table),
                ),
                term::Kind::PresentEvent { event_id, pattern } => {
                    match symbol_table.get_type(event_id) {
                        Typ::SMEvent { .. } => lir::contract::Term::some(
                            lir::contract::Term::ident(symbol_table.get_name(pattern)),
                        ),
                        Typ::SMTimeout { .. } => {
                            lir::contract::Term::some(lir::contract::Term::ok(
                                lir::contract::Term::ident(symbol_table.get_name(pattern)),
                            ))
                        }
                        _ => unreachable!(),
                    }
                }
                term::Kind::TimeoutEvent { event_id } => match symbol_table.get_type(event_id) {
                    Typ::SMTimeout { .. } => lir::contract::Term::some(lir::contract::Term::err()),
                    _ => unreachable!(),
                },
            }
        }
    }
}
