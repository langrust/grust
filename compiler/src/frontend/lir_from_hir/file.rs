use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
