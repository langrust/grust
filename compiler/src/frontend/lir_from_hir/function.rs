prelude! {
    hir::Function,
    lir::{ Block, item::Function as LIRFunction, Stmt },
}

use super::LIRFromHIR;

impl LIRFromHIR for Function {
    type LIR = LIRFunction;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let Function {
            id,
            contract,
            statements,
            returned,
            ..
        } = self;

        // get function name
        let name = symbol_table.get_name(id).clone();

        // get function inputs
        let inputs = symbol_table
            .get_function_input(id)
            .into_iter()
            .map(|id| {
                (
                    symbol_table.get_name(*id).clone(),
                    symbol_table.get_type(*id).clone(),
                )
            })
            .collect::<Vec<_>>();

        // get function output type
        let output = symbol_table.get_function_output_type(id).clone();

        // tranforms into LIR statements
        let mut statements = statements
            .into_iter()
            .map(|statement| statement.lir_from_hir(symbol_table))
            .collect::<Vec<_>>();
        statements.push(Stmt::ExprLast {
            expression: returned.lir_from_hir(symbol_table),
        });

        // transform contract
        let contract = contract.lir_from_hir(symbol_table);

        LIRFunction {
            name,
            inputs,
            output,
            body: Block { statements },
            contract,
        }
    }
}
