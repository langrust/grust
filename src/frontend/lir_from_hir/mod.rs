use crate::{common::r#type::Type, lir::item::import::Import, symbol_table::SymbolTable};

/// LIR file construction from HIR project.
pub mod file;

/// LIR function construction from HIR function.
pub mod function;

/// LIR node files construction from HIR node.
pub mod node;

/// LIR expression construction from HIR expression.
pub mod expression;

/// LIR statement construction from HIR statement.
pub mod statement;

/// LIR expression construction from HIR stream expression.
pub mod stream_expression;

/// LIR node file construction from HIR unitary node.
pub mod unitary_node;

/// LIR item construction from HIR typedef.
pub mod typedef;

/// LIR contract construction from HIR contract.
pub mod contract;

/// LIR memory construction from HIR typedef.
pub mod memory;

/// LIR pattern construction from HIR pattern.
pub mod pattern;

/// LIR type construction from HIR type.
pub mod r#type;

/// HIR transformation into LIR.
pub trait LIRFromHIR {
    /// Corresponding LIR construct.
    type LIR;

    /// Transforms HIR into LIR.
    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR;

    /// Get type from LIR.
    fn get_type(&self) -> Option<&Type> {
        None
    }

    /// Tell if LIR construct is an IfThenElse operator.
    fn is_if_then_else(&self, _symbol_table: &SymbolTable) -> bool {
        false
    }

    /// Get imports from LIR.
    fn get_imports(&self, _symbol_table: &SymbolTable) -> Vec<Import> {
        vec![]
    }
}
