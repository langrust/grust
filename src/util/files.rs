use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}