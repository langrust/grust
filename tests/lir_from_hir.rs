use codespan_reporting::files::SimpleFiles;

use grustine::lir_from_hir;

#[test]
fn lir_from_hir_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/counter.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(counter_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/blinking.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(blinking_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let button_management_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/button_management.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let button_management_condition_match_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/lir_from_hir/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_condition_match_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/lir_from_hir/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_using_function_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/pid.gr").expect("unkown file"),
    );

    let project = lir_from_hir(pid_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}
