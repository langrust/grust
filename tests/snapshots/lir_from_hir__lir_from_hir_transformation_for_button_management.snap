---
source: tests/lir_from_hir.rs
expression: project
---
items:
  - Enumeration:
      name: Button
      elements:
        - Released
        - Pressed
  - Enumeration:
      name: ResetState
      elements:
        - InProgress
        - Confirmed
  - NodeFile:
      name: counter_o
      imports: []
      input:
        node_name: counter_o
        elements:
          - identifier: res
            type: Boolean
          - identifier: inc
            type: Integer
      state:
        node_name: counter_o
        elements:
          - Buffer:
              identifier: mem_o
              type: Integer
        step:
          node_name: counter_o
          output_type: Integer
          body:
            - Let:
                identifier: o
                expression:
                  IfThenElse:
                    condition:
                      InputAccess:
                        identifier: res
                    then_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              Literal:
                                literal:
                                  Integer: 0
                    else_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              FunctionCall:
                                function:
                                  Identifier:
                                    identifier: " + "
                                arguments:
                                  - FunctionCall:
                                      function:
                                        Identifier:
                                          identifier: (_)
                                      arguments:
                                        - MemoryAccess:
                                            identifier: mem_o
                                  - InputAccess:
                                      identifier: inc
          state_elements_step:
            - identifier: mem_o
              expression:
                Identifier:
                  identifier: o
          output_expression:
            Identifier:
              identifier: o
          contract:
            requires: []
            ensures: []
            invariant: []
            assert: []
        init:
          node_name: counter_o
          state_elements_init:
            - BufferInit:
                identifier: mem_o
                initial_value:
                  Integer: 0
          invariant_initialisation: []
  - NodeFile:
      name: reset_button_user_reset_request
      imports:
        - Enumeration: Button
        - NodeFile: counter_o
      input:
        node_name: reset_button_user_reset_request
        elements:
          - identifier: button_state
            type:
              Enumeration: Button
          - identifier: period
            type: Integer
          - identifier: reset_limit_1
            type: Integer
          - identifier: reset_limit_2
            type: Integer
      state:
        node_name: reset_button_user_reset_request
        elements:
          - Buffer:
              identifier: mem_res
              type: Boolean
          - Buffer:
              identifier: mem_user_reset_request
              type: Boolean
          - CalledNode:
              identifier: counter_o_counter
              node_name: counter_o
        step:
          node_name: reset_button_user_reset_request
          output_type: Boolean
          body:
            - Let:
                identifier: res
                expression:
                  FunctionCall:
                    function:
                      Identifier:
                        identifier: " && "
                    arguments:
                      - FunctionCall:
                          function:
                            Identifier:
                              identifier: (_)
                          arguments:
                            - FunctionCall:
                                function:
                                  Identifier:
                                    identifier: " == "
                                arguments:
                                  - InputAccess:
                                      identifier: button_state
                                  - Literal:
                                      literal:
                                        Enumeration:
                                          - Button
                                          - Pressed
                      - FunctionCall:
                          function:
                            Identifier:
                              identifier: (_)
                          arguments:
                            - MemoryAccess:
                                identifier: mem_res
            - Let:
                identifier: counter
                expression:
                  NodeCall:
                    node_identifier: counter_o_counter
                    input_name: CounterOInput
                    input_fields:
                      - - res
                        - Identifier:
                            identifier: res
                      - - inc
                        - InputAccess:
                            identifier: period
            - Let:
                identifier: user_reset_request
                expression:
                  Match:
                    matched:
                      InputAccess:
                        identifier: button_state
                    arms:
                      - - Constant:
                            constant:
                              Enumeration:
                                - Button
                                - Released
                            location: ~
                        - ~
                        - MemoryAccess:
                            identifier: mem_user_reset_request
                      - - Constant:
                            constant:
                              Enumeration:
                                - Button
                                - Pressed
                            location: ~
                        - ~
                        - FunctionCall:
                            function:
                              Identifier:
                                identifier: " >= "
                            arguments:
                              - Identifier:
                                  identifier: counter
                              - FunctionCall:
                                  function:
                                    Identifier:
                                      identifier: " + "
                                  arguments:
                                    - InputAccess:
                                        identifier: reset_limit_1
                                    - InputAccess:
                                        identifier: reset_limit_2
          state_elements_step:
            - identifier: mem_res
              expression:
                FunctionCall:
                  function:
                    Identifier:
                      identifier: " == "
                  arguments:
                    - InputAccess:
                        identifier: button_state
                    - Literal:
                        literal:
                          Enumeration:
                            - Button
                            - Released
            - identifier: mem_user_reset_request
              expression:
                Identifier:
                  identifier: user_reset_request
          output_expression:
            Identifier:
              identifier: user_reset_request
          contract:
            requires: []
            ensures: []
            invariant: []
            assert: []
        init:
          node_name: reset_button_user_reset_request
          state_elements_init:
            - BufferInit:
                identifier: mem_res
                initial_value:
                  Boolean: true
            - BufferInit:
                identifier: mem_user_reset_request
                initial_value:
                  Boolean: false
            - CalledNodeInit:
                identifier: counter_o_counter
                node_name: counter_o
          invariant_initialisation: []
  - NodeFile:
      name: reset_button_user_reset_request_state
      imports:
        - Enumeration: Button
        - NodeFile: counter_o
        - Enumeration: ResetState
      input:
        node_name: reset_button_user_reset_request_state
        elements:
          - identifier: button_state
            type:
              Enumeration: Button
          - identifier: period
            type: Integer
          - identifier: reset_limit_1
            type: Integer
      state:
        node_name: reset_button_user_reset_request_state
        elements:
          - Buffer:
              identifier: mem_res
              type: Boolean
          - CalledNode:
              identifier: counter_o_counter
              node_name: counter_o
        step:
          node_name: reset_button_user_reset_request_state
          output_type:
            Enumeration: ResetState
          body:
            - Let:
                identifier: res
                expression:
                  FunctionCall:
                    function:
                      Identifier:
                        identifier: " && "
                    arguments:
                      - FunctionCall:
                          function:
                            Identifier:
                              identifier: (_)
                          arguments:
                            - FunctionCall:
                                function:
                                  Identifier:
                                    identifier: " == "
                                arguments:
                                  - InputAccess:
                                      identifier: button_state
                                  - Literal:
                                      literal:
                                        Enumeration:
                                          - Button
                                          - Pressed
                      - FunctionCall:
                          function:
                            Identifier:
                              identifier: (_)
                          arguments:
                            - MemoryAccess:
                                identifier: mem_res
            - Let:
                identifier: counter
                expression:
                  NodeCall:
                    node_identifier: counter_o_counter
                    input_name: CounterOInput
                    input_fields:
                      - - res
                        - Identifier:
                            identifier: res
                      - - inc
                        - InputAccess:
                            identifier: period
            - Let:
                identifier: user_reset_request_state
                expression:
                  Match:
                    matched:
                      InputAccess:
                        identifier: button_state
                    arms:
                      - - Constant:
                            constant:
                              Enumeration:
                                - Button
                                - Released
                            location: ~
                        - ~
                        - Literal:
                            literal:
                              Enumeration:
                                - ResetState
                                - Confirmed
                      - - Constant:
                            constant:
                              Enumeration:
                                - Button
                                - Pressed
                            location: ~
                        - ~
                        - IfThenElse:
                            condition:
                              FunctionCall:
                                function:
                                  Identifier:
                                    identifier: " >= "
                                arguments:
                                  - Identifier:
                                      identifier: counter
                                  - InputAccess:
                                      identifier: reset_limit_1
                            then_branch:
                              statements:
                                - ExpressionLast:
                                    expression:
                                      Literal:
                                        literal:
                                          Enumeration:
                                            - ResetState
                                            - InProgress
                            else_branch:
                              statements:
                                - ExpressionLast:
                                    expression:
                                      Literal:
                                        literal:
                                          Enumeration:
                                            - ResetState
                                            - Confirmed
          state_elements_step:
            - identifier: mem_res
              expression:
                FunctionCall:
                  function:
                    Identifier:
                      identifier: " == "
                  arguments:
                    - InputAccess:
                        identifier: button_state
                    - Literal:
                        literal:
                          Enumeration:
                            - Button
                            - Released
          output_expression:
            Identifier:
              identifier: user_reset_request_state
          contract:
            requires: []
            ensures: []
            invariant: []
            assert: []
        init:
          node_name: reset_button_user_reset_request_state
          state_elements_init:
            - BufferInit:
                identifier: mem_res
                initial_value:
                  Boolean: true
            - CalledNodeInit:
                identifier: counter_o_counter
                node_name: counter_o
          invariant_initialisation: []

